packages/flutter/lib/src/services/spell_check.dart

```diff
@@ -0,0 +1,214 @@
+/// results returned by a spell checker.
+@immutable
+class SpellCheckResults {
+  /// Creates results based off those received by spell checking some text input.
+  const SpellCheckResults(this.spellCheckedText, this.suggestionSpans)
+      : assert(spellCheckedText != null),
+        assert(suggestionSpans != null);
+
+  /// The text that the [suggestionSpans] correspond to.
+  final String spellCheckedText;
+
+  /// The spell check results of the [spellCheckedText].
+  ///
+  /// See also:
+  ///
+  ///  * [SuggestionSpan], the ranges of misspelled text and corresponding
+  ///    replacement suggestions.
+  final List<SuggestionSpan> suggestionSpans;
+
+  @override
+  bool operator ==(Object other) {
+    if (identical(this, other)) {
+        return true;
+    }
+
+    return other is SpellCheckResults &&
+        other.spellCheckedText == spellCheckedText &&
+        listEquals<SuggestionSpan>(other.suggestionSpans, suggestionSpans);
+  }
+
+  @override
+  int get hashCode => Object.hash(spellCheckedText, Object.hashAll(suggestionSpans));
+}
+
+/// Determines how spell check results are received for text input.
+abstract class SpellCheckService {
+  /// Facilitates a spell check request.
+  ///
+  /// Returns a [Future] that resolves with a [List] of [SuggestionSpan]s for
+  /// all misspelled words in [text] for the given [locale].
+  Future<List<SuggestionSpan>?> fetchSpellCheckSuggestions(
+    Locale locale, String text,
+  );
+}
+
+/// The service used by default to fetch spell check results for text input.
+///
+/// Any widget may use this service to spell check text by calling
+/// `fetchSpellCheckSuggestions(locale, text)` with an instance of this class.
+/// This is currently only supported by Android.
+///
+/// See also:
+///
+///  * [SpellCheckService], the service that this implements and may be
+///    overriden for use by [EditableText].
+///  * [EditableText], which may use this service to fetch results.
+class DefaultSpellCheckService implements SpellCheckService {
+  /// Creates service to spell check text input by default via communcication
+  /// over the spell check [MethodChannel].
+  DefaultSpellCheckService() {
+    spellCheckChannel = SystemChannels.spellCheck;
+  }
+
+  /// The last recieved [SuggestionSpan]s from the shell side.
+  List<SuggestionSpan>? lastSavedSpans;
+
+  /// The text corresponding to the [lastSavedSpans].
+  String? lastSavedText;
+
+  /// The channel used to communicate with the shell side to complete spell
+  /// check requests.
+  late MethodChannel spellCheckChannel;
+
+  /// Merges two lists of spell check [SuggestionSpan]s.
+  ///
+  /// Used in cases where the text has not changed, but the spell check results
+  /// received from the shell side have. This case is caused by IMEs (GBoard,
+  /// for instance) that ignore the composing region when spell checking text.
+  static List<SuggestionSpan> mergeResults(
```

Comment:
It looks like this method assumes that the results lists will be sorted by range start.  Is that documented anywhere?